### PR TITLE
feat: added better custom header support

### LINKED ISSUE #1346: [Bug]: Homepage ignores the headers set in `next.config.js` when deployed to Netlify
### Summary

I set some headers in `next.config.js` as it's recommended on Next.js docs. You also mentioned the version 4 of this cool plugin has full support of Next.js headers, rewrites, and redirects on the release page, so I expect all the headers to be set, but It didn't happen for me. Netlify says this plugin is used and everything else works fine. Also, I tried setting `NODE_ENV=production` environment and redeploying, but got no luck.

### Steps to reproduce

1. Set headers in `next.config.js`
2. Push changes to GitHub
3. Netlify builds and deploys the project automatically
4. Check the headers of the website using an online tool

### A link to a reproduction repository

_No response_

### Plugin version

4.7.0

### More information about your build

- [ ] I am building using the CLI
- [X] I am building using file-based configuration (`netlify.toml`)

### What OS are you using?

Windows

### Your netlify.toml file

<details>
  <summary>`netlify.toml`</summary>
  
```toml
[context.production.environment]
NEXT_PUBLIC_X="X"

[context.staging.environment]
NEXT_PUBLIC_X="X"
```

</details>

### Your public/_redirects file

<details>
  <summary>`_redirects`</summary>
  
```toml
# Paste content of your `_redirects` file here
```

</details>

### Your `next.config.js` file

<details>
  <summary>`next.config.js`</summary>
  
```toml
const { withSentryConfig } = require('@sentry/nextjs');
const nextBundleAnalyzer = require('@next/bundle-analyzer');
const { i18n } = require('./next-i18next.config');

const securityHeaders = [
  {
    key: 'X-Content-Type-Options',
    value: 'nosniff',
  },
  {
    key: 'Content-Security-Policy',
    value: "default-src 'self';",
  },
  {
    key: 'Strict-Transport-Security',
    value: 'max-age=63072000; includeSubDomains; preload',
  },
  {
    key: 'Strict-Transport-Security',
    value: 'max-age=63072000; includeSubDomains; preload',
  },
  {
    key: 'Cross-Origin-Opener-Policy',
    value: 'same-origin',
  },
  {
    key: 'Cross-Origin-Resource-Policy',
    value: 'same-origin',
  },
];

/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack(config, { webpack }) {
    config.module.rules.push({
      test: /\.svg$/,
      issuer: /\.tsx?$/,
      use: [
        {
          loader: '@svgr/webpack',
          options: {
            typescript: true,
            filenameCase: 'kebab',
            memo: true,
          },
        },
      ],
    });

    if (process.env.NODE_ENV === 'production') {
      config.plugins.push(
        new webpack.DefinePlugin({
          __SENTRY_DEBUG__: false,
        }),
      );
    }

    return config;
  },
  reactStrictMode: true,
  swcMinify: true,
  compiler: {
    removeConsole: process.env.NODE_ENV === 'production',
  },
  pageExtensions: [
    'page.tsx',
  ],
  async rewrites() {
    const rewrites = [
      { source: '/s/:slug*', destination: '/p/:slug*' },
      {
        source: '/api/:slug*',
        destination: 'backend_url:slug*',
      },
    ];

    return rewrites;
  },
  async redirects() {
    return [
      {
        source: '/p/:slug*',
        destination: '/s/:slug*',
        permanent: true,
      },
    ];
  },
  async headers() {
    if (process.env.NODE_ENV === 'production') {
      return [
        {
          source: '/:path*',
          headers: securityHeaders,
        },
      ];
    }

    return [];
  },
  i18n,
};

const withBundleAnalyzer = nextBundleAnalyzer({
  enabled: process.env.ANALYZE === 'true',
});

module.exports = withBundleAnalyzer(withSentryConfig(nextConfig));
```

</details>

### Builds logs (or link to your logs)

<details>
  <summary>Build logs</summary>
  
```
# Paste logs here
```

</details>

### Function logs

<details>
  <summary>Function logs</summary>
  
```
# Paste logs here
```

</details>

### .next JSON files

<details>
  <summary>generated .next JSON files</summary>
  
```
# Paste file contents here. Please check there isn't any private info in them
# You can either build locally, or download the deploy from Netlify by clicking the arrow next to the deploy time.
```

</details>

### PR DESCRIPTION
**Pull request by nickytonline**
### Summary

Improves support for [Next.js custom headers](https://nextjs.org/docs/api-reference/next.config.js/headers). Currently the `headers()` function in next.config.js only works for [SSR](https://nextjs.org/docs/basic-features/pages#server-side-rendering)/[ISR](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration). [Headers in netlify.toml](https://docs.netlify.com/configure-builds/file-based-configuration/#headers) or a [_headers file](https://docs.netlify.com/routing/headers/#syntax-for-the-headers-file) are completely ignored for SSR/ISR. For [SSG pages](https://nextjs.org/docs/basic-features/pages#static-generation-recommended), only the headers defined in netlify.toml or a _headers file are recognized and headers defined in `headers()` in next.config.js are ignored.

Previously, you had to manually copy from the `headers()` function to netlify.toml or vice-versa so all headers were applied. Now all that is required is defining custom headers in the `headers()` function of next.config.js. The build plugin will update the Netlify configuration at build time with all the custom headers defined in the `headers()` function of next.config.js.

You can still add headers in netlify.toml or _headers, but those additional headers will only apply to SSG pages.

@KyleBlankRollins, we'll need a documentation update for this as well. Happy to write this up, just not sure how we usually work with the documentation team.

Note that ISR and SSR are not affected by this plugin fix.

### Test plan

The source code for test site can be found at https://github.com/nickytonline/next-plugin-issue-1346

Note that unless you load a page explicitly from the browser address bar, Next.js client-side routing will kick in so as you navigate to a new page, from the first-page load, you will not see the custom headers in subsequent pages that load from the client-side as the browser is not refreshing. You will need to refresh the browser to see the headers.

If you want to test the fix locally, you'll need to build the fix locally and run `npm link`, then in the test site repository, you'll need to run `npm link @netlify/plugin-nextjs`. Just a reminder that the fix can only be tested in a site that was deployed to Netlify using the plugin fix. There is no way to test this locally.

This is the configuration being used in the sites deployed in the test scenarios, https://github.com/nickytonline/next-plugin-issue-1346/blob/96471d10bb7be58287859e1e62af227eacf9e127/next.config.js

1. First check with the existing plugin that headers aren't being served. Navigate to https://62991c0c8fa1390082a46ace--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/.
2. If you inspect the network tab for the request of the page rendered, you'll notice none of the custom headers I created are there, e.g. the `x-all-pages-header` header
3. Navigate to the deployed site that used the plugin fix, https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/.
4. Open the Network tab and you should see the following headers:

```
x-all-pages-header: all pages header value
x-all-pages-header2: all pages header value2
x-header-for-root-of-site: this is the header for the root of the site
x-header-for-root-of-site2: this is another the header for the root of the site
```

Network tab showing custom headers being served from the Netlify configuration

5. Navigate to https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/en/some-path/yolo
6. Open the Network tab and you should see the following headers:

```
x-all-locales-header-some-path: all locales header some path value
x-all-pages-header: all pages header value
x-all-pages-header2: all pages header value2
x-en-locale-specific-yolo-header: en locale specific yolo header value
```

Network tab showing custom headers being served from the Netlify configuration6

7. Navigate to https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/en/some-path and https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/es/some-path
8. Open the Network tab and you should see the following headers for each of those pages:

```
x-all-locales-header-some-path: all locales header some path value
x-all-pages-header: all pages header value
x-all-pages-header2: all pages header value2
```

Network tab showing custom headers being served from the Netlify configuration

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal

Closes #1346

### Standard checks:

- [x] Check the Deploy Preview's Demo site for your PR's functionality
- [ ] Add docs when necessary

---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.

A player heading a goal in a soccer match

### CODE REVIEW

File: plugin/src/helpers/config.ts
Comment (nickytonline) [inline plugin/src/helpers/config.ts:52]: I expose this through `getNextConfig(...)` so it's easily accessible, but so far it's only used in the `onPostBuild` step of the plugin. Originally, I had that same code in the `onPostBuild` instead of encapsulating it in `getNextConfig(...)`.

File: test/index.js
Comment (nickytonline) [inline test/index.js:254]: Prettier in action

**Comment by nickytonline** [comment]
The failing tests are complaining about not finding a routes-manifest.json file, but I check if it exists before loading it. Will take a peek at this in the AM.

File: plugin/src/helpers/config.ts
Comment (ascorbic) [inline plugin/src/helpers/config.ts:16]: [pebble] You can use [the full type](https://github.com/netlify/netlify-plugin-nextjs/blob/main/plugin/src/helpers/types.ts#L42)

File: plugin/src/helpers/config.ts
Comment (ascorbic) [inline plugin/src/helpers/config.ts:50]: Are you sure this is needed? I don't think the routes manifest can be anywhere except `join(publish, 'routes-manifest.json')`

File: plugin/src/helpers/config.ts
Comment (ascorbic) [inline plugin/src/helpers/config.ts:141]: [pebble] I'm not 100% sure the Next.js path pattern is exactly compatible with ours. Take a look at the equivalent for redirects

**Review by ascorbic — COMMENTED** [review]
This is looking great. I think this approach is exactly the right one. Just a couple of comments and questions

File: plugin/src/helpers/config.ts
Comment (nickytonline) [reply plugin/src/helpers/config.ts:16]: Didn't realize there was already a full type defined already. I'll definitely use that instead.

File: test/index.js
Comment (ericapisani) [inline test/index.js:798]: [sand] it doesn't look like the `async` keyword is needed here and in the test below, could this  be removed?

**Comment by KyleBlankRollins** [comment]
Thanks for tagging me, @nickytonline! You have several options for a docs update. I'll list them in order of how long it'll likely take to complete (shortest time to longest time).
- Create a PR in the docs repo and update/add content as necessary to the [Next.js page](https://docs.netlify.com/integrations/frameworks/next-js/). I'll review it and suggest changes/edits.
- Create an issue in the docs repo. I'll pick up the issue when I can and create my own PR, then tag you for review.
- Wait for me to create an issue, then pick it up and create the PR and tag you for review.
I mention this because I'll be slower than usual this week.

**Comment by nickytonline** [comment]
> Thanks for tagging me, @nickytonline! You have several options for a docs update. I'll list them in order of how long it'll likely take to complete (shortest time to longest time).
> 
> * Create a PR in the docs repo and update/add content as necessary to the [Next.js page](https://docs.netlify.com/integrations/frameworks/next-js/). I'll review it and suggest changes/edits.
> * Create an issue in the docs repo. I'll pick up the issue when I can and create my own PR, then tag you for review.
> * Wait for me to create an issue, then pick it up and create the PR and tag you for review.
>   I mention this because I'll be slower than usual this week.

Thanks Kyle! I'll definitely create a PR to the docs repo for this.

**Comment by nickytonline** [comment]
TODO Add docs related to the PR

File: plugin/src/helpers/config.ts
Comment (nickytonline) [reply plugin/src/helpers/config.ts:136]: @ascorbic, based on e.g,

netlify.toml

```
[build]
command = "yarn build"
publish = ".next"

[[plugins]]
package = "@netlify/plugin-nextjs"
```

`headers()` function in next.config.js

```javascript
/** @type {import('next').NextConfig} */
const nextConfig = {
  async headers() {
    return [
      {
        source: '/',
        headers: [
          {
            key: 'x-hello',
            value: 'yo',
          },
          {
            key: 'x-hello2',
            value: 'yo',
          },
        ],
      },
      {
        source: '/:path*',
        headers: [
          {
            key: 'x-hello',
            value: 'yo path star',
          },
          {
            key: 'x-hello2',
            value: 'yo path star again',
          },
        ],
      },
      {
        source: '/yolo',
        headers: [
          {
            key: 'x-turtle',
            value: 'snapping',
          },
        ],
      },
    ]
  },
};

module.exports = nextConfig
```
and the generated routes-manifest.json

```json
...
  "headers": [
    {
      "source": "/",
      "headers": [
        {
          "key": "x-hello",
          "value": "yo"
        },
        {
          "key": "x-hello2",
          "value": "yo"
        }
      ],
      "regex": "^/(?:/)?$"
    },
    {
      "source": "/:path*",
      "headers": [
        {
          "key": "x-hello",
          "value": "yo path star"
        },
        {
          "key": "x-hello2",
          "value": "yo path star again"
        }
      ],
      "regex": "^(?:/((?:[^/]+?)(?:/(?:[^/]+?))*))?(?:/)?$"
    },
    {
      "source": "/yolo",
      "headers": [
        {
          "key": "x-turtle",
          "value": "snapping"
        }
      ],
      "regex": "^/yolo(?:/)?$"
    }
  ],
  ...
  ```

on post-build, the Netlify configuration is updated to

```
────────────────────────────────────────────────────────────────
  5. @netlify/plugin-nextjs (onPostBuild event)                 
────────────────────────────────────────────────────────────────

Next.js cache saved.
Netlify configuration property "headers" value changed to [
  { for: '/', values: { 'x-hello': 'yo', 'x-hello2': 'yo' } },
  {
    for: '/:path*',
    values: { 'x-hello': 'yo path star', 'x-hello2': 'yo path star again' }
  },
  { for: '/yolo', values: { 'x-turtle': 'snapping' } }
].
```

If they happen to have redirects in the netlify.toml already, on post-build appends to them

```
────────────────────────────────────────────────────────────────
  5. @netlify/plugin-nextjs (onPostBuild event)                 
────────────────────────────────────────────────────────────────

Next.js cache saved.
Netlify configuration property "headers" value changed to [
  { for: '/*', values: { 'Some-Key': 'Some Value' } },
  { for: '/', values: { 'Some-Key2': 'Some Value Just Slash' } },
  { for: '/', values: { 'x-hello': 'yo', 'x-hello2': 'yo' } },
  {
    for: '/:path*',
    values: { 'x-hello': 'yo path star', 'x-hello2': 'yo path star again' }
  },
  { for: '/yolo', values: { 'x-turtle': 'snapping' } }
].
```

**Comment by ascorbic** [comment]
@nickytonline I don't think `/:path*` is a valid pattern for us. We don't support named splats like that. It would need to be `/*`.

**Comment by nickytonline** [comment]
> @nickytonline I don't think `/:path*` is a valid pattern for us. We don't support named splats like that. It would need to be `/*`.

@ascorbic, based on https://docs.netlify.com/routing/headers/#syntax-for-the-headers-file I think we're good?

> Paths can contain * or :placeholders. A :placeholder matches anything except /, while a * matches anything.

**Review by ericapisani — APPROVED** [review]
LGTM

**Comment by ascorbic** [comment]
I am checking with runtime to be sure though

File: plugin/src/helpers/config.ts
Comment (ascorbic) [inline plugin/src/helpers/config.ts:137]:
```suggestion
      for: source.replace(/:[^*]+\*$/g, '*'),
```
We should only match it at the end of the path

File: plugin/src/helpers/config.ts
Comment (ascorbic) [inline plugin/src/helpers/config.ts:168]:
```suggestion
    // Explicitly checking false to make the check simpler.
```

File: plugin/src/helpers/config.ts
Comment (ericapisani) [inline plugin/src/helpers/config.ts:171]:
```suggestion
    // Explicitly checking false to make the check simpler.
```

File: plugin/src/helpers/config.ts
Comment (ericapisani) [inline plugin/src/helpers/config.ts:197]: [sand] It's small enough that it's not something that necessarily needs to be addressed now, but this code looks like it could be broken into its own helper function because it deals entirely with local specific path when that functionality is enabled.

File: test/index.js
Comment (ericapisani) [inline test/index.js:818]: [pebble] something to address in future work but I think we need to start breaking this test file down into different files

File: test/index.js
Comment (ericapisani) [inline test/index.js:819]: [question/potential boulder]: given this function will be used in the `onPostBuild` step of the plugin, is there any particular reason why a test (or an assertion within an existing test for `onPostBuild`) doesn't exist to confirm that `generateCustomHeaders` is called within that step? 

Doing so will ensure that the functionality is working in the area where we expect it to, and the more rigorous testing of the functionality can remain here.

File: test/index.js
Comment (nickytonline) [reply test/index.js:818]: Yeah, that occurred to me as well, but I thought I'd wait after we wrap up all our convos we've been having about testing in general.

File: test/index.js
Comment (ericapisani) [inline test/index.js:724]: [boulder] this test and it's contents appear twice in this file

**Review by ericapisani — COMMENTED** [review]
Other than the one comment, this is LGTM. Will approve once that's cleaned up

File: test/index.js
Comment (nickytonline) [reply test/index.js:724]: Oof. How'd that get in there? 😬

File: plugin/src/helpers/config.ts
Comment (nickytonline) [reply plugin/src/helpers/config.ts:197]: I'm going to hold off on this for now.

**Review by ericapisani — APPROVED** [review]

**Pull request merged**
